node/src/pruner.rs

```diff
@@ -0,0 +1,407 @@
+    #[error(transparent)]
+    P2p(#[from] P2pError),
+
+    /// An error propagated from the [`Store`] module.
+    #[error(transparent)]
+    Store(#[from] StoreError),
+
+    #[error(transparent)]
+    Blockstore(#[from] blockstore::Error),
+}
+
+pub struct Pruner {
+    cancellation_token: CancellationToken,
+}
+
+pub struct PrunerArgs<S, B>
+where
+    S: Store,
+    B: Blockstore,
+{
+    /// Headers storage.
+    pub store: Arc<S>,
+    /// Block storage.
+    pub blockstore: Arc<B>,
+    /// Event publisher.
+    pub event_pub: EventPublisher,
+}
+
+impl Pruner {
+    pub fn start<S, B>(args: PrunerArgs<S, B>) -> Self
+    where
+        S: Store + 'static,
+        B: Blockstore + 'static,
+    {
+        let cancellation_token = CancellationToken::new();
+        let event_pub = args.event_pub.clone();
+
+        let mut worker = Worker::new(args, cancellation_token.child_token());
+
+        spawn(async move {
+            if let Err(e) = worker.run().await {
+                error!("Pruner stopped because of a fatal error: {e}");
+
+                event_pub.send(NodeEvent::FatalPrunerError {
+                    error: e.to_string(),
+                });
+            }
+        });
+
+        Pruner { cancellation_token }
+    }
+
+    pub fn stop(&self) {
+        self.cancellation_token.cancel();
+    }
+}
+
+impl Drop for Pruner {
+    fn drop(&mut self) {
+        self.cancellation_token.cancel();
+    }
+}
+
+struct Worker<S, B>
+where
+    S: Store + 'static,
+    B: Blockstore + 'static,
+{
+    cancellation_token: CancellationToken,
+    event_pub: EventPublisher,
+    store: Arc<S>,
+    blockstore: Arc<B>,
+}
+
+impl<S, B> Worker<S, B>
+where
+    S: Store,
+    B: Blockstore,
+{
+    fn new(args: PrunerArgs<S, B>, cancellation_token: CancellationToken) -> Self {
+        Worker {
+            cancellation_token,
+            event_pub: args.event_pub,
+            store: args.store,
+            blockstore: args.blockstore,
+        }
+    }
+
+    async fn run(&mut self) -> Result<()> {
+        let mut last_reported = None;
+        let mut last_removed = None;
+        loop {
+            let pruning_window_end = Time::now().checked_sub(PRUNING_WINDOW).unwrap_or_else(|| {
+                warn!("underflow when computing pruning window start, defaulting to unix epoch");
+                Time::unix_epoch()
+            });
+
+            while let Some(header) = self.get_tail_header_to_prune(&pruning_window_end).await? {
+                let height = header.height().value();
+
+                let cids = self
+                    .store
+                    .get_sampling_metadata(height)
+                    .await?
+                    .map(|m| m.cids)
+                    .unwrap_or_default();
+                for cid in cids {
+                    self.blockstore.remove(&cid).await?;
+                }
+
+                let removed = self.store.remove_last().await?;
```

Comment:
I know that those invariants are held by the 1h gap between syncing and pruning window but I don't like much that we aren't explicitely removing height for which we wiped cid's from blockstore

Reply:
I don't know if there is better way tho, we shouldn't make ranges discontinuous. Can we maybe convert the debug_assert below to hard error?

Reply:
Yeah, we're silently assuming nobody will be messing with the header's tail other than pruner. Since I think we can rely on the syncing/pruning window difference,(so it shouldn't happen), added an error to detect above and halt pruner when we encounter it.